File: c++/implement_strStr.cpp
/*
Implement strStr().

Returns the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.
 */

/*
 * 
*/

#include "helper.h"

class Solution {
public:
    int strStr(string haystack, string needle) {
        if (haystack.size() < needle.size()) {
            return -1;
        }
        for (int i = 0; i <= haystack.size() - needle.size(); i++) {
            int j = 0;
            for (; j < needle.size() && needle[j] == haystack[i + j]; j++) {

            }
            if (j == needle.size()) {
                return i;
            }
        }
        return -1;
    }
};

int main() {
    Solution s;
    cout << s.strStr("abcd", "ab") << endl;
    cout << s.strStr("abcd", "cd") << endl;
    cout << s.strStr("abcd", "dc") << endl;

    return 0;
}

File: python/implement_strStr.py
'''
Implement strStr().

Returns the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.

Update (2014-11-02):
The signature of the function had been updated to return the index instead of the pointer. If you still see your function signature returns a char * or String, please click the reload button  to reset your code definition.
'''

'''
Two pointers. Note that we only need to check from position 0 to len(haystack) - len(needle) + 1. This will reduce some unnecessary calculations.
'''

class Solution:
    # @param haystack, a string
    # @param needle, a string
    # @return an integer
    def strStr(self, haystack, needle):
        p1 = 0
        p2 = 0
        if len(needle) == 0:
            return 0
        while p1 < len(haystack) - len(needle) + 1:
            if haystack[p1] == needle[0]:
                p2 = p1
                while p2 < len(haystack) and p2 - p1 < len(needle):
                    if haystack[p2] != needle[p2-p1]:
                        break
                    p2 += 1
                if p2 - p1 == len(needle):
                    return p1
            p1 += 1
        return -1

if __name__ == '__main__':
    s = Solution()
    print s.strStr('abcd', 'cd')
